controllers/src/main/scala/cromwell/pipeline/controller/AkkaHttpClient.scala

```diff
@@ -4,67 +4,100 @@ import akka.actor.ActorSystem
 import akka.http.scaladsl.Http
 import akka.http.scaladsl.model.Uri.Query
 import akka.http.scaladsl.model.headers.RawHeader
-import akka.http.scaladsl.model.{ HttpMethods, HttpRequest, HttpResponse, Uri }
+import akka.http.scaladsl.model.{ HttpEntity, HttpMethods, HttpRequest, HttpResponse, Uri }
 import akka.stream.ActorMaterializer
-import cromwell.pipeline.service.{ HttpClient, Response }
+import cromwell.pipeline.service.{ FailureResponseBody, HttpClient, Response, SuccessResponseBody }
+import play.api.libs.json.{ JsValue, Json, Reads, Writes }
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import de.heikoseeberger.akkahttpplayjson.PlayJsonSupport._
 
 import scala.concurrent.duration._
 import scala.concurrent.{ ExecutionContext, Future }
 
 class AkkaHttpClient(implicit actorSystem: ActorSystem, materializer: ActorMaterializer) extends HttpClient {
   private val expirationTime: FiniteDuration = 300.millis
 
-  override def get(url: String, params: Map[String, String] = Map(), headers: Map[String, String] = Map())(
-    implicit ec: ExecutionContext
-  ): Future[Response] = {
+  override def get[B](url: String, params: Map[String, String] = Map(), headers: Map[String, String] = Map())(
+    implicit ec: ExecutionContext,
+    f: Reads[B]
+  ): Future[Response[B]] = {
     val futureResponse: Future[HttpResponse] = Http().singleRequest(
       HttpRequest(method = HttpMethods.GET, uri = Uri(url).withQuery(Query(params))).withHeaders(parseHeaders(headers))
     )
     responsify(futureResponse)
   }
 
-  private def responsify(futureResponse: Future[HttpResponse])(implicit ec: ExecutionContext): Future[Response] =
+  private def responsify[B](
+    futureResponse: Future[HttpResponse]
+  )(implicit ec: ExecutionContext, f: Reads[B]): Future[Response[B]] =
     futureResponse.flatMap { response =>
-      response.entity.toStrict(expirationTime).map { body =>
-        Response(
-          response.status.intValue(),
-          body.data.utf8String,
-          response.headers.foldLeft(Map[String, List[String]]()) {
-            case (acc, header) =>
-              acc + (header.name() -> (header.value() :: acc.getOrElse(header.name(), Nil)))
-          }
-        )
+      if (response.status.isSuccess()) {
+        successResponse(response)
+      } else {
+        failureResponse(response)
       }
     }
 
+  private def successResponse[B](
+    response: HttpResponse
+  )(implicit ec: ExecutionContext, f: Reads[B]): Future[Response[B]] =
+    Unmarshal(response.entity).to[B].map { b =>
+      Response[B](
+        response.status.intValue(),
+        SuccessResponseBody(b),
+        parseHeaders(response)
+      )
+    }
+
+  private def failureResponse[B](response: HttpResponse)(implicit ec: ExecutionContext): Future[Response[B]] =
+    response.entity.toStrict(expirationTime).map { body =>
+      val pb = parseResponseBody(body)
```

Comment:
I think we cannot assume that the failed response will return a `json`.  Therefore, you should not parse this body to a `JsValue`.